### PR TITLE
coding guidelines: comply with MISRA C:2012 Rule 14.3 and 2.1

### PR DESCRIPTION
**Pull request by Abramo-Bagnara**

### CODE REVIEW

File: drivers/pcie/host/msi.c [hunk 1]
```diff
@@ -268,18 +270,18 @@ bool pcie_msi_enable(pcie_bdf_t bdf,
 
 	base = pcie_get_cap(bdf, PCI_CAP_ID_MSI);
 
-	if (IS_ENABLED(CONFIG_PCIE_MSI_X)) {
```
Comment (erwango) [inline drivers/pcie/host/msi.c:271]: Do I understand correctly that use of `IS_ENABLED` is forbidden by the rules mentioned in title ?
If this is true, a mail to devel mailinglist would be welcome

Comment (Abramo-Bagnara) [reply drivers/pcie/host/msi.c:271]: MISRA Rule 14.3 asks to avoid compile time known condition.
MISRA Rule 2.1 asks (among other things) to avoid unreachable branches (thus very strictly related to the above rule).

In the rare cases where a macro has a conditional on an expression that might be a compile time known condition or not, I've introduced and used a macro called ``CONSTEXPR`` to mean "I'm aware that condition might be constant, it is deliberate".

Where instead the use of a ``#if``, ``COND_CODE_1``, ``IF_ENABLED`` is feasible I've replaced the runtime condition with a preprocessor conditional.

This does not mean that in all cases ``IS_ENABLED`` is forbidden, but only that when code can be improved avoiding to possibly embark unreachable code and/or to put inadvertently constant conditions in code, it is best to take this opportunity and obtain MISRA compliance.

Comment (erwango) [reply drivers/pcie/host/msi.c:271]: So, do I understand correctly that it should not be used on `CONFIG_FOO` anymore ?
In that case, please fix the macro doc which is all about use with CONFIG symbols:
```
/**
 * @brief Check for macro definition in compiler-visible expressions
 *
 * This trick was pioneered in Linux as the config_enabled() macro. It
 * has the effect of taking a macro value that may be defined to "1"
 * or may not be defined at all and turning it into a literal
 * expression that can be handled by the C compiler instead of just
 * the preprocessor. It is often used with a @p CONFIG_FOO macro which
 * may be defined to 1 via Kconfig, or left undefined.
 *
 * That is, it works similarly to <tt>\#if defined(CONFIG_FOO)</tt>
 * except that its expansion is a C expression. Thus, much <tt>\#ifdef</tt>
 * usage can be replaced with equivalents like:
 *
 *     if (IS_ENABLED(CONFIG_FOO)) {
 *             do_something_with_foo
 *     }
 *
 * This is cleaner since the compiler can generate errors and warnings
 * for @p do_something_with_foo even when @p CONFIG_FOO is undefined.
``` 

and then provide a clear example in which is could be used.
Otherwise you'll have to come back on this endlessly.

**Review by erwango — CHANGES_REQUESTED** [review]

File: drivers/pcie/host/msi.c (hunk 1 above)
Comment (Abramo-Bagnara) [reply drivers/pcie/host/msi.c:271]: Although I personally think that the benefits to reduce the compiled/possibly embarked code outweigh by far the convenience to have a configured out branch verified for the lack of syntax errors by the compiler (that is a very poor kind of "testing"), I don't think to have the authority to mandate that.
I think that the steering committee should take an official decision about that, before I can change such comment.

Comment (erwango) [reply drivers/pcie/host/msi.c:271]: @nashif what do you think ?

**Review by jhedberg — CHANGES_REQUESTED** [review]
I have a feeling that the project needs to approve an exception for `IS_ENABLED()`, since it has many advantages over `#ifdef` and should be preferred whenever possible.

**Comment by Abramo-Bagnara** [comment]
The first thing I'd like to clarify is that IS_ENABLED() is a not problem per se, but only when it leads to unreachable code.
The absence of unreachable code is fundamental for DO 178 B/C, ISO 26262, IEC 61508 and has clear implications on coverage, security and code quality in general.
If your proposal is to have an exception for code unreachable due to IS_ENABLED() I think it is very important to list all the advantages of that so to ensure they outweigh what I've written above.

**Comment by jhedberg** [comment]
@Abramo-Bagnara the main point of `IS_ENABLED()` is specifically unreachable code, i.e. exposing all build-time conditional branches to the compiler at once. It's not supposed to replace proper testing of the those branches, but lets you get early warnings/errors of issues without having to do multiple compilations with different Kconfig configurations. It also helps improve code readability, by exposing the reader to conditional expressions written in just one language (C) rather than two (pre-processor & C) which get interpreted at different build stages. I hope we can agree on that C-files littered with #ifdefs all over the place are less readable (which also leads to more bugs and worse maintainability) than those that have less #ifdefs.

**Comment by jhedberg** [comment]
@Abramo-Bagnara also wanted to add, that when I say "unreachable" this is strictly about the C compiler perspective. It's certainly not unreachable from the Zephyr project perspective since `IS_ENABLED()` is supposed to be used only with Kconfig symbols, which, if they are valid, can be enabled.

**Comment by npitre** [comment]
> The first thing I'd like to clarify is that IS_ENABLED() is a not problem per
> se, but only when it leads to unreachable code.

But that's exactly the _point_ of `IS_ENABLED()`: to make large portion
of the code unreachable so that the compiler can optimize it away at
compile time while still allowing for analysis of the code for correctness
by the compiler irrespective of the configuration value (which is not
possible with `#ifdef`.

When an `#ifdef` is used, the enclosed code is not subjected to compiler
oversight. That makes less frequently used options more susceptible to
breakage whereas `IS_ENABLED()` makes many issues visible right away even
if the code is not used in the final build. Furthermore `#ifdef`'s
are uglier and more distracting.

> The absence of unreachable code is fundamental for DO 178 B/C,
> ISO 26262, IEC 61508 and has clear implications on coverage, security
> and code quality in general.

This is madness. Common sense must always prevail over any bureaucratic
rules that show their age. And `IS_ENABLED()` has proven its usefullness
with way more exposure for far more time outside of Zephyr already than any
standard committee could ever dream of.

**Comment by Abramo-Bagnara** [comment]
> When an `#ifdef` is used, the enclosed code is not subjected to compiler
> oversight. That makes less frequently used options more susceptible to
> breakage whereas `IS_ENABLED()` makes many issues visible right away even
> if the code is not used in the final build.

To hear that the check for parser error alone is something valuable sound rather worrisome for my ears, but perhaps it is only me...

> Furthermore #ifdef's are uglier and more distracting.

I disagree on that: as I wrote above the ``#ifdef`` that gives alternative implementations for an abstract functionality depending on config macros is very nice, readable and verifiable.

OTOH ``#ifdef`` in code are not so pretty, but however better than to have unreachable code (also in the hope that their ugliness can stimulate to convert them to nicer abstractions now missing).

About functional standards: it is rather pointless to dismiss them as insane when a buyer ask for compliance to it, I guess this is not going to work with an assessor.

**Comment by nordic-krch** [comment]
I don't understand this. If rule states `Controlling expressions shall not be invariant` then expression like `if (IS_ENABLED(CONFIG_FEATURE))` is not invariant because the code inside is intended to be compiled in certain configurations supported by the project. Source has no control what CONFIG_FEATURE is since it comes from included file. It can be `#define CONFIG_FEATURE runtime_flag`.
In all examples I have found there is a constant which results from that source file, meaning that there is no possible configuration that would lead to that code compilation.

As for benefits of using IS_ENABLED and whole family of runtime expressions which can be evaluated at compile time one example: Logging is relying on it and thanks to that variables used in the log message is evaluated by the compiler. So if you would have variable created for logging message and logging macro would be an ifdef then that would also need to be ifdefed:
```
#if LOCAL_LOG_LEVEL >= LOG_LEVEL_INF
int tmp_freq = x / 1000;
#endif
LOG_INF("freq %d", tmp_freq);
```
Another thing around logging is that it is likely that not all logging levels are complied by CI so if that would be a macro evaluated by preprocessor only when log level is enabled then it is likely that it would generate warnings due to wrong format specifier, undefined variable etc. I've seen that before in the projects where preprocessor was favored over compile time evaluation.

Summing up, huge NO for interpreting IS_ENABLED() as invariant expression and thus non compliant with 14.3 rule.

**Comment by Abramo-Bagnara** [comment]
> As for benefits of using IS_ENABLED and whole family of runtime expressions which can be evaluated at compile time one example: Logging is relying on it and thanks to that variables used in the log message is evaluated by the compiler. So if you would have variable created for logging message and logging macro would be an ifdef then that would also need to be ifdefed:

Currently this is not true, the use of a macro called UNEVALUATED ensures that also when logging is disable format specifiers are checked and involved variables are referenced. Not that I like so much to have a bunch of things computed and never really used, I'd have preferred:

```
#define IF_LOG(stmt) IF_ENABLED(CONFIG_LOG, stmt)

void foo(void) {
  IF_LOG({
      int freq = get_freq();
      LOG_INF("freq %d", freq);
  })
}
```

The good things of that approach is a better abstraction, no unreachable code/unwanted calls and also the fact that when compliance is not wanted the following can be used:

```
#define IF_LOG(stmt) if (IS_ENABLED(CONFIG_LOG)) stmt
```

About the definition of invariant guards we cannot stretch that: a guard is invariant in a **specific** CONFIG settings and not for all of them.

Of course it is possible to deviate justifying that (and keeping tracking thousands of violations), but the PR shows that another way is feasible (0 violations) and that it can lead to introduce better abstractions.

You can see the alternative here https://www.misra.org.uk/app/uploads/2021/06/MISRA-C-2012-Permits-First-Edition.pdf at page 10, but please read carefully the requirements: they come with a cost.

**Comment by Abramo-Bagnara** [comment]
> Are you referring to those requirements?

Yes.

> ad 1. Use of `IS_ENABLED()` is quite good identification.

Also the related reports for unreachable code should be recorded in compliance document.
